Remove `join_workers` from Pipeline destructor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53433 Remove `join_workers` from Pipeline destructor.**

As described in https://github.com/pytorch/pytorch/issues/53413, the
pipeline destructor ends up hanging sometimes. The reason for this is that Pipe
uses daemon threads and as a result these threads could be destroyed before the
Pipe destructor is done. The Pipe destructor then calls `join_workers` which
waits on signals from the worker threads, which might be already dead and
results in the main thread blocking forever.

To resolve this issue, in this PR we remove `join_workers` completely since it
is not necessary to wait for daemon threads.

#Closes: https://github.com/pytorch/pytorch/issues/53413

Differential Revision: [D26863321](https://our.internmc.facebook.com/intern/diff/D26863321/)

pritamdamania committed:
Remove `join_workers` from Pipeline destructor.

As described in https://github.com/pytorch/pytorch/issues/53413, the
pipeline destructor ends up hanging sometimes. The reason for this is that Pipe
uses daemon threads and as a result these threads could be destroyed before the
Pipe destructor is done. The Pipe destructor then calls `join_workers` which
waits on signals from the worker threads, which might be already dead and
results in the main thread blocking forever.

To resolve this issue, in this PR we remove `join_workers` completely since it
is not necessary to wait for daemon threads.

#Closes: https://github.com/pytorch/pytorch/issues/53413

Differential Revision: [D26863321](https://our.internmc.facebook.com/intern/diff/D26863321/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
There are some flake errors but seem unrelated

pritamdamania committed:
Update on "Remove `join_workers` from Pipeline destructor."


As described in https://github.com/pytorch/pytorch/issues/53413, the
pipeline destructor ends up hanging sometimes. The reason for this is that Pipe
uses daemon threads and as a result these threads could be destroyed before the
Pipe destructor is done. The Pipe destructor then calls `join_workers` which
waits on signals from the worker threads, which might be already dead and
results in the main thread blocking forever.

To resolve this issue, in this PR we remove `join_workers` completely since it
is not necessary to wait for daemon threads.

#Closes: https://github.com/pytorch/pytorch/issues/53413

Differential Revision: [D26863321](https://our.internmc.facebook.com/intern/diff/D26863321/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Remove `join_workers` from Pipeline destructor."


As described in https://github.com/pytorch/pytorch/issues/53413, the
pipeline destructor ends up hanging sometimes. The reason for this is that Pipe
uses daemon threads and as a result these threads could be destroyed before the
Pipe destructor is done. The Pipe destructor then calls `join_workers` which
waits on signals from the worker threads, which might be already dead and
results in the main thread blocking forever.

To resolve this issue, in this PR we remove `join_workers` completely since it
is not necessary to wait for daemon threads.

#Closes: https://github.com/pytorch/pytorch/issues/53413

Differential Revision: [D26863321](https://our.internmc.facebook.com/intern/diff/D26863321/)

[ghstack-poisoned]
